### DISCUSSION TITLE
Proposal: introduce operator to inline expressions without relying on inference.

### DISCUSSION THREAD

**Discussion by willnationsdev** [dotnet/csharplang#6842] [Language Ideas] [answered]
## Problem

With LINQ and EF Core, Expressions are seeing ever-increasing usage. The cases where they are most frequent, namely, in method and extension method parameters, allow the compiler to easily infer the fact that the passed-in expression should be taken as-is/literally based on the parameter type.

```cs
// Just an example
public static async Task HasAsync<TContext, TEntity>(this TContext context, Expression<Func<TEntity, bool>> predicate)
    where TContext : DbContext
    where TEntity : class
{
    return await context.Set<TEntity>().AnyAsync(predicate);
}

// Usage. Implicitly interpreted as an Expression rather than as a Func.
var johnExists = await db.HasAsync((MyUser user) => user.NormalizedUserName == "JOHN_SMITH");
```

However, in situations where you are attempting to inline values in an evaluated expression, this inference fails to work.

```cs
var addTenFunc = (int x) => x + 10;
Expression<Func<int, int>> addTenFuncExpr = (int x) => x + 10; // explicit typing is required
```

In F#, disambiguating between functions and expressions for local variable inference is made easy via the `<@ ... @>` operators.

```fs
let addTenFunc = fun x -> x + 10
let addTenFuncExpr = <@ fun x -> x + 10 @>
```

However, that is specific to FSharpExpressions and does not work the same way as System.Linq.Expressions as far as I know.

## Proposal

Drawing from the Razor approach of flipping between C# with `@` and HTML content with `@:`, we could introduce a verbatim-assignment operator (`@=`) such that the following expression is interpreted "verbatim" as a System.Linq.Expression.

```cs
var addTenFunc = (int x) => x + 10;
var addTenFuncExpr = @= (int x) => x + 10;
var addTenFuncExprMultiLine = @= (int x) => {
    return x + 10;
};
var _ = value switch {
    null => @= () => default, // this arm infers the return type based on the following arm, but does so as an Expression.
    _ => @= () => (int?)value, // inferred to be a `Expression<Func<int?>>`, regardless of `value`'s actual type.
};
```

**Comment by svick** [comment, marked as answer]
I'm not convinced new syntax is required here, since [you can do the same thing using a simple function](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSB0AMgJZQCO+AogB4AOYApgM5PED2UTA3Oujfc1YcAPHACswgCoAabJIBKzCABtgAPjXYGdMFNkKlqtQAp02c9n6MW7KKIky5ipivWbtAmxwCU2ALzuOoK2PGjoAG4AhmDYkQAmcZIMUOJW/lo6xsakwNjUvgF52ADU2EiY3qHo8YnJqTr4ACIQALa0xpVAA) (a very similar method is part of LINQKit):

```c#
Expression<Func<T, TResult>> expr<T, TResult>(Expression<Func<T, TResult>> expression) => expression;

var addTenFuncExpr = expr((int x) => x + 10);
```

**Reply by willnationsdev** [reply]
Ah, I see. If it's truly that simple, then maybe it really IS not necessary.
